h/services/annotation_write.py

```diff
@@ -77,31 +77,99 @@ def create_annotation(self, data: dict) -> Annotation:
+        self._validate_group(
+            annotation, enforce_write_permission=enforce_write_permission
+        )
+
+        if (
+            document := data.get("document", {})
+        ) or annotation.target_uri != initial_target_uri:
+            annotation.document = update_document_metadata(
+                self._db,
+                annotation.target_uri,
+                document.get("document_meta_dicts", {}),
+                document.get("document_uri_dicts", {}),
+                updated=annotation.updated,
+            )
+
+        # The search index service by default does not reindex if the existing ES
+        # entry's timestamp matches the DB timestamp. If we're not changing this
+        # timestamp, we need to force reindexing.
+        # pylint: disable=protected-access
+        self._search_index_service._queue.add_by_id(
+            annotation.id,
+            tag=reindex_tag,
+            schedule_in=60,
+            force=not update_timestamp,
+        )
+
+        return annotation
+
+    @staticmethod
+    def _update_annotation_values(annotation: Annotation, data: dict):
+        for key, value in data.items():
+            # Don't set complex things
+            if key in ("document", "extra"):
+                continue
+
+            setattr(annotation, key, value)
+
+        extra = data.get("extra", {})
+        annotation.extra.update(extra)
+
+    def _validate_group(self, annotation: Annotation, enforce_write_permission=True):
         group = annotation.group
         if not group:
             raise ValidationError(
                 "group: " + _(f"Invalid group id {annotation.groupid}")
             )
 
-        # The user must have permission to create an annotation in the group
-        # they've asked to create one in.
-        if not self._has_permission(
+        # The user must have permission to write to the group
+        if enforce_write_permission and not self._has_permission(
             Permission.Group.WRITE, context=GroupContext(annotation.group)
         ):
             raise ValidationError(
                 "group: " + _("You may not create annotations in the specified group!")
             )
 
-        # If no scopes are present, or if the group is configured to allow
-        # annotations outside its scope, there's nothing to do here
-        if not group.scopes or not group.enforce_scope:
-            return
-
-        # The target URI must match at least one
-        # of a group's defined scopes, if the group has any
-        if not url_in_scope(
-            annotation.target_uri, [scope.scope for scope in group.scopes]
+        if (
+            group.scopes  # If we have scopes
+            and group.enforce_scope  # ... and we are enforcing them
+            # the target URI must match one of a group's defined scopes
+            and not url_in_scope(
+                annotation.target_uri, [scope.scope for scope in group.scopes]
+            )
```

Comment:
Small un-related refactor to point out these are two parts of the same condition.